Source/stash.cpp

```diff
@@ -0,0 +1,839 @@
+std::optional<CelSprite> StashNavButtonArt;
+std::optional<CelSprite> StashGoldButtonArt;
+
+void StashDrawSlotBack(const Surface &out, Point targetPosition, Size size)
+{
+	SDL_Rect srcRect = MakeSdlRect(0, 0, size.width, size.height);
+	out.Clip(&srcRect, &targetPosition);
+	if (size.width <= 0 || size.height <= 0)
+		return;
+
+	std::uint8_t *dst = &out[targetPosition];
+	const auto dstPitch = out.pitch();
+
+	for (int hgt = size.height; hgt != 0; hgt--, dst -= dstPitch + size.width) {
+		for (int wdt = size.width; wdt != 0; wdt--) {
+			std::uint8_t pix = *dst;
+			if (pix >= PAL16_BLUE) {
+				if (pix <= PAL16_BLUE + 15)
+					pix -= PAL16_BLUE - PAL16_BEIGE;
+				else if (pix >= PAL16_GRAY)
+					pix -= PAL16_GRAY - PAL16_BEIGE;
+			}
+			*dst++ = pix;
+		}
+	}
+}
+
+/**
+ * @brief Adds an item to a player's StashGrid array
+ * @param stashGridIndex Item's position in StashGrid (this should be the item's topleft grid tile)
+ * @param stashListIndex The item's StashList index (it's expected this already has +1 added to it since StashGrid can't store a 0 index)
+ * @param itemSize Size of item
+ */
+void AddItemToStashGrid(int stashGridIndex, int stashListIndex, Size itemSize)
+{
+	const int pitch = 10;
+	for (int y = 0; y < itemSize.height; y++) {
+		for (int x = 0; x < itemSize.width; x++) {
+			if (x == 0 && y == itemSize.height - 1)
+				Stash.stashGrids[StashPage][stashGridIndex + x] = stashListIndex;
+			else
+				Stash.stashGrids[StashPage][stashGridIndex + x] = -stashListIndex;
+		}
+		stashGridIndex += pitch;
+	}
+}
+
+/**
+ * @brief Gets the size, in stash cells, of the given item.
+ * @param item The item whose size is to be determined.
+ * @return The size, in stash cells, of the item.
+ */
+Size GetStashSize(const Item &item)
+{
+	int itemSizeIndex = item._iCurs + CURSOR_FIRSTITEM;
+	auto size = GetInvItemSize(itemSizeIndex);
+
+	return { size.width / InventorySlotSizeInPixels.width, size.height / InventorySlotSizeInPixels.height };
+}
+
+/**
+ * @brief Checks whether the given item can fit in a belt slot (i.e. the item's size in inventory cells is 1x1).
+ * @param item The item to be checked.
+ * @return 'True' in case the item can fit a belt slot and 'False' otherwise.
+ */
+bool FitsInBeltSlot(const Item &item)
+{
+	return GetStashSize(item) == Size { 1, 1 };
+}
+
+/**
+ * @brief Checks whether the given item can be placed on the belt. Takes item size as well as characteristics into account. Items
+ * that cannot be placed on the belt have to be placed in the inventory instead.
+ * @param item The item to be checked.
+ * @return 'True' in case the item can be placed on the belt and 'False' otherwise.
+ */
+bool CanBePlacedOnBelt(const Item &item)
+{
+	return FitsInBeltSlot(item)
+	    && item._itype != ItemType::Gold
+	    && item._iStatFlag
+	    && AllItemsList[item.IDidx].iUsable;
+}
+
+int SwapStashItem(Item *a, Item *b)
+{
+	std::swap(*a, *b);
+
+	return b->_iCurs + CURSOR_FIRSTITEM;
+}

```

Comment:
Any reason these need to be reimplemented instead of using the inv.cpp definitions?